### PR TITLE
global-shortcuts: Try harder to sanitize input options

### PR DESCRIPTION
**Pull request by aleixpol**
Where we are taking arguments as variant, make sure it has the correct types inside.

### CODE REVIEW

**Comment by aleixpol** [comment]
Ping? I thought this was somewhat pressing... @GeorgesStavracas

**Comment by GeorgesStavracas** [comment]
It is, but I can't find time to prioritize portal reviews

File: src/global-shortcuts.c [hunk 1]
```diff
@@ -311,9 +374,35 @@ handle_bind_shortcuts (XdpDbusGlobalShortcuts *object,
...
+                           G_N_ELEMENTS (global_shortcuts_bind_shortcuts_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  options = g_variant_builder_end (&options_builder);
+
+
```
Comment (jadahl) [inline src/global-shortcuts.c:395]: Stray newline.

File: src/global-shortcuts.c [hunk 2]
```diff
@@ -216,10 +221,23 @@ handle_create_session (XdpDbusGlobalShortcuts *object,
   Request *request = request_from_invocation (invocation);
   g_autoptr(GError) error = NULL;
   g_autoptr(XdpDbusImplRequest) impl_request = NULL;
+  g_autoptr(GVariant) options;
```
Comment (jadahl) [inline src/global-shortcuts.c:224]: Must always initialize autoptrs to `NULL`

File: src/global-shortcuts.c [hunk 3]
```diff
@@ -216,10 +221,23 @@ handle_create_session (XdpDbusGlobalShortcuts *object,
...
   Session *session;
 
   REQUEST_AUTOLOCK (request);
 
+  GVariantBuilder options_builder;
```
Comment (jadahl) [inline src/global-shortcuts.c:229]: Declarations in the top block (applies here and there below).

**Review by jadahl — CHANGES_REQUESTED** [review]

File: src/global-shortcuts.c [hunk 4]
```diff
@@ -410,8 +499,23 @@ handle_list_shortcuts (XdpDbusGlobalShortcuts *object,
...
   g_autoptr(GError) error = NULL;
 
+  g_autoptr(GVariant) options;
+
   REQUEST_AUTOLOCK (request);
 
+  GVariantBuilder options_builder;
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options, &options_builder,
+                           global_shortcuts_bind_shortcuts_options, //list has the same options as bind
```
[the remark this answers is not in the document]
Comment (aleixpol) [reply src/global-shortcuts.c:509]: G_N_ELEMENTS doesn't seem to like it.

[the remark this answers is not in the document]
Comment (jadahl) [reply src/global-shortcuts.c:506]: Maybe just copy the list then? It's not unthinkable that they will eventually diverge.

[the remark this answers is not in the document]
Comment (aleixpol) [reply src/global-shortcuts.c:506]: Sure, if it makes you happier. :)

File: src/global-shortcuts.c [hunk 5]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+gboolean
+xdp_verify_shortcuts (GVariant *shortcuts,
+                      GVariantBuilder *filtered,
+                      GError **error)
+{
+  gboolean ret = TRUE;
+  gchar *shortcut_name;
+  GVariant *values;
+
+  g_autoptr(GVariantIter) iter = g_variant_iter_new (shortcuts);
```
Comment (jadahl) [inline src/global-shortcuts.c:334]: A couple..

File: src/global-shortcuts.c [hunk 6]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+      if (shortcut_name[0] == 0)
+        {
+          if (*error == NULL)
+            g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                          "Unexpected empty shortcut id");
+          ret = FALSE;
+          continue;
+        }
+
+      GVariantBuilder shortcut_builder;
```
Comment (jadahl) [inline src/global-shortcuts.c:346]: .. of more places with ..

File: src/global-shortcuts.c [hunk 7]
```diff
@@ -311,9 +373,34 @@ handle_bind_shortcuts (XdpDbusGlobalShortcuts *object,
   Session *session;
   g_autoptr(XdpDbusImplRequest) impl_request = NULL;
   g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) options = NULL;
+  g_autoptr(GVariant) shortcuts = NULL;
 
   REQUEST_AUTOLOCK (request);
 
+  GVariantBuilder options_builder;
```
Comment (jadahl) [inline src/global-shortcuts.c:381]: ... declarations ...

File: src/global-shortcuts.c [hunk 8]
```diff
@@ -311,9 +373,34 @@ handle_bind_shortcuts (XdpDbusGlobalShortcuts *object,
...
+                           G_N_ELEMENTS (global_shortcuts_bind_shortcuts_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  options = g_variant_builder_end (&options_builder);
+
+  GVariantBuilder shortcuts_builder;
```
Comment (jadahl) [inline src/global-shortcuts.c:394]: ... outside of the initial block.

File: src/global-shortcuts.c [hunk 9]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+      GVariantBuilder shortcut_builder;
+      g_variant_builder_init (&shortcut_builder, G_VARIANT_TYPE_VARDICT);
+      if (!xdp_filter_options (values, &shortcut_builder,
+                              global_shortcuts_keys,
+                              G_N_ELEMENTS (global_shortcuts_keys),
+                              error))
+        {
+          ret = FALSE;
+        }
+      g_variant_builder_add(filtered, "(sa{sv})", g_variant_builder_end (&shortcut_builder));
```
Comment (jadahl) [inline src/global-shortcuts.c:355]: Missing space before `(`.

File: src/global-shortcuts.c [hunk 10]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+  gboolean ret = TRUE;
+  gchar *shortcut_name;
+  GVariant *values;
+
+  g_autoptr(GVariantIter) iter = g_variant_iter_new (shortcuts);
+  while (g_variant_iter_loop (iter, "sa{sv}", &shortcut_name, &values))
+    {
+      if (shortcut_name[0] == 0)
+        {
+          if (*error == NULL)
```
Comment (jadahl) [inline src/global-shortcuts.c:339]: This `NULL` heck is already handled by `g_set_error()`.

File: src/global-shortcuts.c [hunk 11]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
     }
 }
 
+static XdpOptionKey global_shortcuts_keys[] = {
+  { "description", G_VARIANT_TYPE_STRING, NULL },
+  { "preferred_trigger", G_VARIANT_TYPE_STRING, NULL },
+};
+
+gboolean
```
Comment (jadahl) [inline src/global-shortcuts.c:325]: `static gboolean`

File: src/global-shortcuts.c [hunk 12]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+
+  g_autoptr(GVariantIter) iter = g_variant_iter_new (shortcuts);
+  while (g_variant_iter_loop (iter, "sa{sv}", &shortcut_name, &values))
+    {
+      if (shortcut_name[0] == 0)
+        {
+          if (*error == NULL)
+            g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                          "Unexpected empty shortcut id");
+          ret = FALSE;
```
Comment (jadahl) [inline src/global-shortcuts.c:342]: Can't we just `return FALSE;` here?

File: src/global-shortcuts.c [hunk 13]
```diff
@@ -299,6 +317,50 @@ shortcuts_bound_cb (GObject *source_object,
...
+        }
+
+      GVariantBuilder shortcut_builder;
+      g_variant_builder_init (&shortcut_builder, G_VARIANT_TYPE_VARDICT);
+      if (!xdp_filter_options (values, &shortcut_builder,
+                              global_shortcuts_keys,
+                              G_N_ELEMENTS (global_shortcuts_keys),
+                              error))
+        {
+          ret = FALSE;
```
Comment (jadahl) [inline src/global-shortcuts.c:353]: Same here with a replacing

```c
        {
          ret = FALSE;
        }
```
with
```c
        return FALSE;
```

File: src/global-shortcuts.c [hunk 14]
```diff
@@ -409,9 +500,23 @@ handle_list_shortcuts (XdpDbusGlobalShortcuts *object,
...
 
   REQUEST_AUTOLOCK (request);
 
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options, &options_builder,
+                           global_shortcuts_list_shortcuts_options,
+                           G_N_ELEMENTS (global_shortcuts_list_shortcuts_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
```
Comment (jadahl) [inline src/global-shortcuts.c:514]: This needs an `g_variant_builder_clear()`.

File: src/global-shortcuts.c [hunk 15]
```diff
@@ -311,9 +373,34 @@ handle_bind_shortcuts (XdpDbusGlobalShortcuts *object,
...
+      return G_DBUS_METHOD_INVOCATION_HANDLED;
+    }
+
+  options = g_variant_builder_end (&options_builder);
+
+  GVariantBuilder shortcuts_builder;
+  g_variant_builder_init (&shortcuts_builder, G_VARIANT_TYPE_ARRAY);
+  if (!xdp_verify_shortcuts (arg_shortcuts, &shortcuts_builder,
+                             &error))
+    {
```
Comment (jadahl) [inline src/global-shortcuts.c:398]: This needs a `g_variant_builder_clear();`

File: src/global-shortcuts.c [hunk 16]
```diff
@@ -216,10 +221,23 @@ handle_create_session (XdpDbusGlobalShortcuts *object,
...
 
   REQUEST_AUTOLOCK (request);
 
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  if (!xdp_filter_options (arg_options, &options_builder,
+                           global_shortcuts_create_session_options,
+                           G_N_ELEMENTS (global_shortcuts_create_session_options),
+                           &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
```
Comment (jadahl) [inline src/global-shortcuts.c:236]: This needs a `g_variant_builder_clear();`

**Review by jadahl — CHANGES_REQUESTED** [review]

File: src/global-shortcuts.c (hunk 5 above)
Comment (jadahl) [reply src/global-shortcuts.c:334]: Some `git grep` tells me that is to be the most common way in xdg-desktop-portal.

Comment (aleixpol) [reply src/global-shortcuts.c:335]: This is the only other case where `g_variant_iter_new` is used:
https://github.com/flatpak/xdg-desktop-portal/blob/main/src/background.c#L264

Comment (jadahl) [reply src/global-shortcuts.c:335]: What I checked was `g_autoptr` in general.

Comment (aleixpol) [reply src/global-shortcuts.c]: Sorry, done.

**Pull request merged**